File: src/skills/demand_forecasting.py
"""
DemandForecastingSkill — used by the EventAgent.

Gives the EventAgent tools to score each event's revenue opportunity
based on attendance, event type, time of day, and duration.
The agent uses these scores to rank events before passing them to the Scheduler.
"""

import json
import math
from typing import Any

from .base import Skill


class DemandForecastingSkill(Skill):
    @property
    def name(self) -> str:
        return "demand_forecasting"

    @property
    def description(self) -> str:
        return "Score and rank events by food truck revenue potential."

    @property
    def prompt_module(self) -> str:
        return """
## Demand Forecasting Skill
You have access to demand forecasting tools. Use them to:
- Call `forecast_demand` for every event you discover to get a revenue score.
- Call `score_event_opportunity` to combine demand with practical factors (duration, time of day).
- Only pass events with an opportunity_score >= 40 to the final ranked list.
- Always sort your final answer by opportunity_score descending so the Scheduler
  picks the highest-value events first.
"""

    @property
    def tools(self) -> list[dict]:
        return [
            {
                "name": "forecast_demand",
                "description": (
                    "Forecast food truck customer demand for an event based on its type and attendance. "
                    "Returns estimated customers, average order value, and a raw demand score (0-100)."
                ),
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "event_id": {"type": "string"},
                        "event_category": {
                            "type": "string",
                            "description": "Event type: market, conference, music, food, sports, festival, etc.",
                        },
                        "expected_attendance": {"type": "integer"},
                        "duration_hours": {
                            "type": "number",
                            "description": "How long the event runs in hours",
                        },
                    },
                    "required": ["event_id", "event_category", "expected_attendance", "duration_hours"],
                },
            },
            {
                "name": "score_event_opportunity",
                "description": (
                    "Combine raw demand with time-of-day and event duration to produce a final "
                    "opportunity score (0-100). Higher is better for deployment."
                ),
                "input_schema": {
                    "type": "object",
                    "properties": {
                        "event_id": {"type": "string"},
                        "demand_score": {"type": "number", "description": "Raw demand score from forecast_demand"},
                        "start_hour": {
                            "type": "integer",
                            "description": "Hour of day the event starts (0-23, UTC)",
                        },
                        "duration_hours": {"type": "number"},
                        "estimated_revenue": {"type": "number", "description": "High-end revenue estimate in USD"},
                    },
                    "required": ["event_id", "demand_score", "start_hour", "duration_hours", "estimated_revenue"],
                },
            },
        ]

    def handle_tool_call(self, tool_name: str, tool_input: dict[str, Any]) -> str:
        handlers = {
            "forecast_demand": self._forecast_demand,
            "score_event_opportunity": self._score_event_opportunity,
        }
        handler = handlers.get(tool_name)
        if not handler:
            return json.dumps({"error": f"Unknown tool: {tool_name}"})
        try:
            return json.dumps(handler(**tool_input), default=str)
        except Exception as exc:
            return json.dumps({"error": str(exc)})

    # ------------------------------------------------------------------
    # Mock implementations — replace with ML model / historical data
    # ------------------------------------------------------------------

    # Conversion rates and average order values by event category
    _CATEGORY_PROFILE = {
        "music":      {"conversion": 0.12, "avg_order": 14.0},
        "conference": {"conversion": 0.09, "avg_order": 16.0},
        "market":     {"conversion": 0.14, "avg_order": 11.0},
        "food":       {"conversion": 0.08, "avg_order": 13.0},
        "festival":   {"conversion": 0.15, "avg_order": 12.0},
        "sports":     {"conversion": 0.10, "avg_order": 10.0},
    }
    _DEFAULT_PROFILE = {"conversion": 0.08, "avg_order": 11.0}

    def _forecast_demand(
        self, event_id: str, event_category: str, expected_attendance: int, duration_hours: float
    ) -> dict:
        profile = self._CATEGORY_PROFILE.get(event_category.lower(), self._DEFAULT_PROFILE)
        estimated_customers = int(expected_attendance * profile["conversion"])
        estimated_revenue = round(estimated_customers * profile["avg_order"], 2)

        # Demand score: scale 0-100 based on customers per hour
        customers_per_hour = estimated_customers / max(duration_hours, 1)
        demand_score = min(100, round(customers_per_hour * 2, 1))

        return {
            "event_id": event_id,
            "estimated_customers": estimated_customers,
            "avg_order_value": profile["avg_order"],
            "estimated_revenue": estimated_revenue,
            "demand_score": demand_score,
        }

    def _score_event_opportunity(
        self,
        event_id: str,
        demand_score: float,
        start_hour: int,
        duration_hours: float,
        estimated_revenue: float,
    ) -> dict:
        # Peak meal hours (11-14, 17-21) get a bonus
        if 11 <= start_hour <= 14 or 17 <= start_hour <= 21:
            time_bonus = 15
        elif 9 <= start_hour <= 16:
            time_bonus = 5
        else:
            time_bonus = -10  # late night / early morning penalty

        # Longer events give more serving time (capped at 6h bonus)
        duration_bonus = min(10, (duration_hours - 2) * 2) if duration_hours > 2 else 0

        # Revenue floor bonus — high revenue events get a small push
        revenue_bonus = min(10, math.log10(max(estimated_revenue, 1)) * 2)

        opportunity_score = round(
            min(100, max(0, demand_score + time_bonus + duration_bonus + revenue_bonus)), 1
        )

        return {
            "event_id": event_id,
            "opportunity_score": opportunity_score,
            "breakdown": {
                "demand_score": demand_score,
                "time_bonus": time_bonus,
                "duration_bonus": duration_bonus,
                "revenue_bonus": round(revenue_bonus, 1),
            },
        }

File: supabase/functions/_shared/skills/demand_forecasting.ts
import Anthropic from "npm:@anthropic-ai/sdk";

// ---------------------------------------------------------------------------
// Tool definitions (Anthropic format)
// ---------------------------------------------------------------------------

export const DEMAND_FORECASTING_TOOLS: Anthropic.Tool[] = [
  {
    name: "forecast_demand",
    description:
      "Forecast food truck customer demand for an event based on its type and attendance. " +
      "Returns estimated customers, average order value, and a raw demand score (0-100).",
    input_schema: {
      type: "object" as const,
      properties: {
        event_id: { type: "string" },
        event_category: {
          type: "string",
          description: "Event type: market, conference, music, food, sports, festival, etc.",
        },
        expected_attendance: { type: "integer" },
        duration_hours: {
          type: "number",
          description: "How long the event runs in hours",
        },
      },
      required: ["event_id", "event_category", "expected_attendance", "duration_hours"],
    },
  },
  {
    name: "score_event_opportunity",
    description:
      "Combine raw demand with time-of-day and event duration to produce a final " +
      "opportunity score (0-100). Higher is better for deployment.",
    input_schema: {
      type: "object" as const,
      properties: {
        event_id: { type: "string" },
        demand_score: {
          type: "number",
          description: "Raw demand score from forecast_demand",
        },
        start_hour: {
          type: "integer",
          description: "Hour of day the event starts (0-23, UTC)",
        },
        duration_hours: { type: "number" },
        estimated_revenue: {
          type: "number",
          description: "High-end revenue estimate in USD",
        },
      },
      required: [
        "event_id",
        "demand_score",
        "start_hour",
        "duration_hours",
        "estimated_revenue",
      ],
    },
  },
];

// ---------------------------------------------------------------------------
// Category profiles (direct port from Python)
// ---------------------------------------------------------------------------

interface CategoryProfile {
  conversion: number;
  avg_order: number;
}

const CATEGORY_PROFILE: Record<string, CategoryProfile> = {
  music:      { conversion: 0.12, avg_order: 14.0 },
  conference: { conversion: 0.09, avg_order: 16.0 },
  market:     { conversion: 0.14, avg_order: 11.0 },
  food:       { conversion: 0.08, avg_order: 13.0 },
  festival:   { conversion: 0.15, avg_order: 12.0 },
  sports:     { conversion: 0.10, avg_order: 10.0 },
};

const DEFAULT_PROFILE: CategoryProfile = { conversion: 0.08, avg_order: 11.0 };

// ---------------------------------------------------------------------------
// Tool handlers (pure computation, no async)
// ---------------------------------------------------------------------------

function forecastDemand(input: Record<string, unknown>): Record<string, unknown> {
  const eventId = input["event_id"] as string;
  const category = ((input["event_category"] as string) ?? "").toLowerCase();
  const attendance = input["expected_attendance"] as number;
  const durationHours = input["duration_hours"] as number;

  const profile = CATEGORY_PROFILE[category] ?? DEFAULT_PROFILE;
  const estimatedCustomers = Math.floor(attendance * profile.conversion);
  const estimatedRevenue = Math.round(estimatedCustomers * profile.avg_order * 100) / 100;

  const customersPerHour = estimatedCustomers / Math.max(durationHours, 1);
  const demandScore = Math.min(100, Math.round(customersPerHour * 2 * 10) / 10);

  return {
    event_id: eventId,
    estimated_customers: estimatedCustomers,
    avg_order_value: profile.avg_order,
    estimated_revenue: estimatedRevenue,
    demand_score: demandScore,
  };
}

function scoreEventOpportunity(input: Record<string, unknown>): Record<string, unknown> {
  const eventId = input["event_id"] as string;
  const demandScore = input["demand_score"] as number;
  const startHour = input["start_hour"] as number;
  const durationHours = input["duration_hours"] as number;
  const estimatedRevenue = input["estimated_revenue"] as number;

  // Peak meal hours (11-14, 17-21) get a bonus
  let timeBonus: number;
  if ((startHour >= 11 && startHour <= 14) || (startHour >= 17 && startHour <= 21)) {
    timeBonus = 15;
  } else if (startHour >= 9 && startHour <= 16) {
    timeBonus = 5;
  } else {
    timeBonus = -10; // late night / early morning penalty
  }

  // Longer events give more serving time (capped at 6h bonus = 10 pts)
  const durationBonus = durationHours > 2 ? Math.min(10, (durationHours - 2) * 2) : 0;

  // Revenue floor bonus — high revenue events get a small push
  const revenueBonus = Math.min(10, Math.log10(Math.max(estimatedRevenue, 1)) * 2);

  const rawScore = demandScore + timeBonus + durationBonus + revenueBonus;
  const opportunityScore = Math.round(Math.min(100, Math.max(0, rawScore)) * 10) / 10;

  return {
    event_id: eventId,
    opportunity_score: opportunityScore,
    breakdown: {
      demand_score: demandScore,
      time_bonus: timeBonus,
      duration_bonus: durationBonus,
      revenue_bonus: Math.round(revenueBonus * 10) / 10,
    },
  };
}

export function handleDemandForecastingTool(
  toolName: string,
  input: Record<string, unknown>,
): string {
  try {
    if (toolName === "forecast_demand") {
      return JSON.stringify(forecastDemand(input));
    }
    if (toolName === "score_event_opportunity") {
      return JSON.stringify(scoreEventOpportunity(input));
    }
    return JSON.stringify({ error: `Unknown demand forecasting tool: ${toolName}` });
  } catch (err) {
    return JSON.stringify({ error: String(err) });
  }
}

// ---------------------------------------------------------------------------
// Prompt module injected into EventAgent system prompt
// ---------------------------------------------------------------------------

export const DEMAND_FORECASTING_PROMPT_MODULE = `
## Demand Forecasting Skill
You have access to demand forecasting tools. Use them to:
- Call \`forecast_demand\` for every event you discover to get a revenue score.
- Call \`score_event_opportunity\` to combine demand with practical factors (duration, time of day).
- Only pass events with an opportunity_score >= 40 to the final ranked list.
- Always sort your final answer by opportunity_score descending so the Scheduler
  picks the highest-value events first.
`.trim();
